quinn-proto/src/congestion/cubic.rs

```diff
@@ -0,0 +1,277 @@
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use super::{Controller, ControllerFactory};
+use std::cmp;
+
+/// CUBIC Constants.
+///
+/// These are recommended value in RFC8312.
+const BETA_CUBIC: f64 = 0.7;
+
+const C: f64 = 0.4;
+
+/// CUBIC State Variables.
+///
+/// We need to keep those variables across the connection.
+/// k, w_max, w_last_max is described in the RFC.
+#[derive(Debug, Default, Clone)]
+pub struct State {
+    k: f64,
+
+    w_max: f64,
+
+    w_last_max: f64,
+
+    // Used in CUBIC fix (see on_packet_sent())
+    last_sent_time: Option<Instant>,
+
+    // Store cwnd increment during congestion avoidance.
+    cwnd_inc: u64,
+}
+
+/// CUBIC Functions.
+///
+/// Note that these calculations are based on a count of cwnd as bytes,
+/// not packets.
+/// Unit of t (duration) and RTT are based on seconds (f64).
+impl State {
+    // K = cbrt(w_max * (1 - beta_cubic) / C) (Eq. 2)
+    fn cubic_k(&self, max_datagram_size: u64) -> f64 {
+        let w_max = self.w_max / max_datagram_size as f64;
+        libm::cbrt(w_max * (1.0 - BETA_CUBIC) / C)
```

Comment:
Why `libm` rather than [`f32::cbrt`](https://doc.rust-lang.org/std/primitive.f32.html#method.cbrt)`?

Reply:
that's an effect of [inspiration from quiche's implementation](https://github.com/cloudflare/quiche/blob/master/src/recovery/cubic.rs#L125)

Reply:
looks like it's to support no_std

Reply:
We don't support no_std, so it would be better to avoid the extra dependency.